pkg/kubelet/volume/populator/desired_state_of_world_populator.go

```diff
@@ -119,17 +136,67 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 	for _, volumeToMount := range dswp.desiredStateOfWorld.GetVolumesToMount() {
 		if _, podExists :=
-			dswp.podManager.GetPodByUID(volumeToMount.Pod.UID); !podExists {
-			glog.V(10).Infof(
-				"Removing volume %q (volSpec=%q) for pod %q from desired state.",
-				volumeToMount.VolumeName,
-				volumeToMount.VolumeSpec.Name(),
-				volumeToMount.PodName)
-
-			dswp.desiredStateOfWorld.DeletePodFromVolume(
-				volumeToMount.PodName, volumeToMount.VolumeName)
-			dswp.deleteProcessedPod(volumeToMount.PodName)
+			dswp.podManager.GetPodByUID(volumeToMount.Pod.UID); podExists {
+			continue
 		}
+
+		dswp.timeOfLastGetPodStatus = time.Now()
+
+		// Once a pod has been deleted from kubelet pod manager, do not
+		// delete it from volume manager. Instead call out to the kubelet
+		// containerRuntime interface to verify all containers in the pod
+		// have been terminated.
+		podStatus, podStatusErr := dswp.kubeContainerRuntime.GetPodStatus(
+			volumeToMount.Pod.UID,
+			volumeToMount.Pod.Name,
+			volumeToMount.Pod.Namespace)
+		if podStatusErr != nil {
+			glog.Errorf(
+				"yosaad: Pod %q/%q (%q) has been removed from pod manager. However, kubeContainerRuntime.GetPodStatus returned error %v. Therefore it will not be removed from volume manager.",
+				volumeToMount.Pod.Namespace,
+				volumeToMount.Pod.Name,
+				volumeToMount.Pod.UID,
+				podStatusErr)
+			continue
+		} else {
```

Comment:
nit: remove the `else {` block now that you skip the rest of the iteration in line 160.